### PR TITLE
BPE support.

### PR DESCRIPTION
**Pull request by stephenroller**
This could be more memory efficient by passing a specialized format (pseudofile of word <tab> count) into the learn_bpe. The API calls would be more complex. For now, submitting functional version that works fine.

If we need to run this on something with an extremely large training data set, we could run into problems.

We could make it *much* more memory efficient and cleaner API by reimplementing or forking the subword-nmt repo, but I think we probably want to stick to the official version.

### CODE REVIEW

File: parlai/core/dict.py [hunk 1]
```diff
@@ -129,6 +136,12 @@ def add_cmdline_args(argparser):
             dictionary.add_argument(
                 '--dict-lower', default=DictionaryAgent.default_lower, type='bool',
                 help='Whether or not to lowercase all text seen.')
+            dictionary.add_argument(
```
Comment (jaseweston) [inline parlai/core/dict.py]: how do you turn off bpe?

Comment (stephenroller) [reply parlai/core/dict.py]: BPE is only enabled if you pass `--dict-tokenizer bpe` as an argument.

File: parlai/core/dict.py [hunk 2]
```diff
@@ -365,6 +387,28 @@ def tokenize(self, text, building=False):
...
+        Finalize and freeze the dictionary.
+
+        If using BPE tokenization, performs the codec learning.
+        """
+
+        if self.tokenizer != 'bpe':
+            # only BPE needs the second pass
+            return
+
+        # Now that we have a second pass,
```
Comment (stephenroller) [inline parlai/core/dict.py]: should update this comment.

File: parlai/core/dict.py [hunk 3]
```diff
@@ -365,6 +387,31 @@ def tokenize(self, text, building=False):
...
+        Finalize and freeze the dictionary.
+
+        If using BPE tokenization, performs the codec learning.
+        """
+
+        if self.tokenizer != 'bpe':
+            # only BPE needs the second pass
+            return
+
+        # Let the BPE model learn its codecs, then use the encodings to to
```
Comment (klshuster) [inline parlai/core/dict.py]: nit: extra "to"

**Comment by stephenroller** [comment]
Fixed the comment bug. Will merge after CI fix.

**Pull request merged**